README.md
improved\_backoff
=================

[![image](https://img.shields.io/badge/python-3.7-blue.svg)](https://www.python.org/downloads/release/python-370)
[![image](https://img.shields.io/badge/python-3.8-blue.svg)](https://www.python.org/downloads/release/python-380)
[![image](https://img.shields.io/badge/python-3.9-blue.svg)](https://www.python.org/downloads/release/python-390)
[![image](https://img.shields.io/badge/python-3.10-blue.svg)](https://www.python.org/downloads/release/python-3100)

[![image](https://github.com/Kirusi/improved_backoff/workflows/tests/badge.svg)](https://github.com/Kirusi/improved_backoff/actions/workflows/tests.yml)
[![image](https://kirusi.github.io/improved_backoff/coverage.svg)](https://github.com/Kirusi/improved_backoff/actions/workflows/coverage.yml)
[![image](https://img.shields.io/pypi/v/improved_backoff.svg)](https://pypi.python.org/pypi/improved_backoff)
[![image](https://img.shields.io/github/license/kirusi/improved_backoff)](https://github.com/kirusi/improved_backoff/blob/master/LICENSE)

**Function decoration for backoff and retry**

This is a fork of an excellent Python library
[backoff](https://github.com/litl/backoff). The library was forked from
version 2.2.1 (October 5, 2022) This version includes 2 PRs proposed in
the original repo:

-   [Correct check for max\_time
    parameter](https://github.com/litl/backoff/pull/130)
-   [Using \"timeit\" module for time
    management](https://github.com/litl/backoff/pull/185)

In order to use this module import it under `backoff` alias and use it
the same way as the original module

```python
import improved_backoff as backoff

@backoff.on_exception(backoff.expo, requests.exceptions.RequestException)
def get_url(url):
    return requests.get(url)
```


CHANGELOG.md
# Changelog

## [v2.2.2] - 2023-06-23

- Fix loggers duplication https://github.com/litl/backoff/issues/179

## [v2.2.1] - 2022-10-05

- Fix type hint for wait generators https://github.com/litl/backoff/issues/177

## [v2.2.0] - 2022-10-05

- Improve type annotation for wait generators from @hauntsaninja
- Include exception in details dict of call handlers from @petamas
- Improve type hints for call handlers from @cdce8p

## [v2.1.2] - 2022-06-08
### Fixed

- Don't use importlib.metadata for __version__ https://github.com/litl/backoff/issues/166

## [v2.1.1] - 2022-06-08
### Fixed

- Fix bug with max_tries/max_time callables https://github.com/litl/backoff/issues/164

## [v2.1.0] - 2022-06-07
### Changed

- Get max_tries/max_time values for every call fixes #160 (from @jvrsantacruz)

## [v2.0.1] - 2022-04-27
### Changed
- Allow None for jitter keyword arg (typing)


## [v2.0.0] - 2022-04-26
### Added
- Add raise_on_giveup keyword arg for decorators
- Add backoff.runtime wait generator for dynamically setting wait times based
  on target function return value or exception details
### Changed
- Improve type hints for on_success, on_backoff, on_giveup handlers
- Use decorator-specific detail and handler type hints
- Optionally use typing_extensions for python 3.7 type hinting
- Drop python 3.6 support
- Add python 3.10 support

## [v1.11.1] - 2021-07-14
### Fixed
- Update __version__ in backoff module

## [v1.11.0] - 2021-07-12
### Added
- Configurable logging levels for backoff and giveup events
### Changed
- Minor documentation fixes

## [v1.10.0] - 2019-12-07
### Changed
- Allow sync decorator call from async function
- NOTE: THIS WILL BE THE FINAL PYTHON 2.7 COMPATIBLE RELEASE.

## [v1.9.2] - 2019-11-19
### Changed
- Don't include tests and changelog in distribution

## [v1.9.1] - 2019-11-18
### Changed
- Include tests and changelog in distribution

## [v1.9.0] - 2019-11-16
### Changed
- Support python 3.8

## [v1.8.1] - 2019-10-11
### Changed
- Use arguments in log messages rather than fully formatting log
  https://github.com/litl/backoff/pull/82 from @lbernick

## [v1.8.0] - 2018-12-20
### Added
- Custom loggers
- Iterable intervals for constant wait_gen for predefined wait sequences
### Changed
- Give up on StopIteration raised in wait generators
- Nullary jitter signature deprecation warning

## [v1.7.0] - 2018-11-23
### Changed
- Support Python 3.7
- Drop support for async in Python 3.4
- Drop support for Python 2.6
- Update development dependencies
- Use poetry for dependencies and packaging

## [v1.6.0] - 2018-07-14
### Changed
- Change default log level from ERROR to INFO
- Log retries on exception as INFO

## [v1.5.0] - 2018-04-11
### Added
- Add max_time keyword argument

## [v1.4.3] - 2017-05-22
### Changed
- Add license to source distribution

## [v1.4.2] - 2017-04-25
### Changed
- Use documented logger name https://github.com/litl/backoff/pull/32
  from @pquentin

## [v1.4.1] - 2017-04-21
### Added
- Expose __version__ at package root
### Changed
- Fix checking for running sync version in coroutine in case when event
  loop is not set from @rutsky

## [v1.4.0] - 2017-02-05
### Added
- Async support via `asyncio` coroutines (Python 3.4) from @rutsky
### Changed
- Refactor `backoff` module into package with identical API

## [v1.3.2] - 2016-11-18
### Changed
- Don't log retried args and kwargs by default
- README.rst syntax highlighting from @dethi

## [v1.3.1] - 2016-08-08
### Changed
- Include README.rst in source distribution (fixes package)

## [v1.3.0] - 2016-08-08
### Added
- Support runtime configuration with optional callable kwargs
- Add giveup kwarg for exception inspection
### Changed
- Documentation fixes

## [v1.2.1] - 2016-05-27
### Changed
- Documentation fixes


## [v1.2.0] - 2016-05-26
### Added
- 'Full jitter' algorithm from @jonascheng

### Changed
- Jitter function now accepts raw value and returns jittered value
- Change README to reST for the benefit of pypi :(
- Remove docstring doc generation and make README canonical

## [v1.1.0] - 2015-12-08
### Added
- Event handling for success, backoff, and giveup
- Change log
### Changed
- Docs and test for multi exception invocations
- Update dev environment test dependencies

## [v1.0.7] - 2015-02-10
### Changed
- Fix string formatting for python 2.6

## [v1.0.6] - 2015-02-10
### Added
- Coveralls.io integration from @singingwolfboy
### Changed
- Fix logging bug for function calls with tuple params

## [v1.0.5] - 2015-02-03
### Changed
- Add a default interval of 1 second for the constant generator
- Improve on_predicate stop condition avoiding extra sleep

## [v1.0.4] - 2014-08-12
### Added
- Python 2.6 support from @Bonko
- Python 3.0 support from @robyoung
- Run tests in Travis from @robyoung

## [v1.0.3] - 2014-06-05
### Changed
- Make logging unicode safe
- Log on_predicate backoff as INFO rather than ERROR


backoff.py
# coding:utf-8
"""
Function decoration for backoff and retry

This module provides function decorators which can be used to wrap a
function such that it will be retried until some condition is met. It
is meant to be of use when accessing unreliable resources with the
potential for intermittent failures i.e. network resources and external
APIs. Somewhat more generally, it may also be of use for dynamically
polling resources for externally generated content.

For examples and full documentation see the README at
https://github.com/litl/backoff
"""
from __future__ import unicode_literals

import functools
import operator
import logging
import random
import time
import traceback
import sys


# Use module-specific logger with a default null handler.
logger = logging.getLogger(__name__)

if sys.version_info < (2, 7, 0):  # pragma: no cover
    class NullHandler(logging.Handler):
        def emit(self, record):
            pass
    logger.addHandler(NullHandler())
else:
    logger.addHandler(logging.NullHandler())  # pragma: no cover

logger.setLevel(logging.ERROR)


def expo(base=2, factor=1, max_value=None):
    """Generator for exponential decay.

    Args:
        base: The mathematical base of the exponentiation operation
        factor: Factor to multiply the exponentation by.
        max_value: The maximum value to yield. Once the value in the
             true exponential sequence exceeds this, the value
             of max_value will forever after be yielded.
    """
    n = 0
    while True:
        a = factor * base ** n
        if max_value is None or a < max_value:
            yield a
            n += 1
        else:
            yield max_value


def fibo(max_value=None):
    """Generator for fibonaccial decay.

    Args:
        max_value: The maximum value to yield. Once the value in the
             true fibonacci sequence exceeds this, the value
             of max_value will forever after be yielded.
    """
    a = 1
    b = 1
    while True:
        if max_value is None or a < max_value:
            yield a
            a, b = b, a + b
        else:
            yield max_value


def constant(interval=1):
    """Generator for constant intervals.

    Args:
        interval: The constant value in seconds to yield.
    """
    while True:
        yield interval


def factor(wait_gen, f):
    """Wrapped decay by factor.

    Args:
        wait_gen: A generator yielding successive wait times in
            seconds.
        factor: Factor to multiply the wait_gen by.
    """
    @functools.wraps(wait_gen)
    def deco(*args, **kwargs):
        wait = wait_gen(*args, **kwargs)
        while True:
            yield next(wait) * f
    return deco


def random_jitter(value):
    """Jitter the value a random number of milliseconds.

    This adds up to 1 second of additional time to the original value.
    Prior to backoff version 1.2 this was the default jitter behavior.

    Args:
        value: The unadulterated backoff value.
    """
    return value + random.random()


def full_jitter(value):
    """Jitter the value across the full range (0 to value).

    This corresponds to the "Full Jitter" algorithm specified in the
    AWS blog's post on the performance of various jitter algorithms.
    (http://www.awsarchitectureblog.com/2015/03/backoff.html)

    Args:
        value: The unadulterated backoff value.
    """
    return random.uniform(0, value)


def on_predicate(wait_gen,
                 predicate=operator.not_,
                 max_tries=None,
                 jitter=full_jitter,
                 on_success=None,
                 on_backoff=None,
                 on_giveup=None,
                 **wait_gen_kwargs):
    """Returns decorator for backoff and retry triggered by predicate.

    Args:
        wait_gen: A generator yielding successive wait times in
            seconds.
        predicate: A function which when called on the return value of
            the target function will trigger backoff when considered
            truthily. If not specified, the default behavior is to
            backoff on falsey return values.
        max_tries: The maximum number of attempts to make before giving
            up. In the case of failure, the result of the last attempt
            will be returned.  The default value of None means their
            is no limit to the number of tries.
        jitter: A function of the value yielded by wait_gen returning
            the actual time to wait. This distributes wait times
            stochastically in order to avoid timing collisions across
            concurrent clients. Wait times are jittered by default
            using the full_jitter function. Jittering may be disabled
            altogether by passing jitter=None.
        on_success: Callable (or iterable of callables) with a unary
            signature to be called in the event of success. The
            parameter is a dict containing details about the invocation.
        on_backoff: Callable (or iterable of callables) with a unary
            signature to be called in the event of a backoff. The
            parameter is a dict containing details about the invocation.
        on_giveup: Callable (or iterable of callables) with a unary
            signature to be called in the event that max_tries
            is exceeded.  The parameter is a dict containing details
            about the invocation.
        **wait_gen_kwargs: Any additional keyword args specified will be
            passed to wait_gen when it is initialized.  Any callable
            args will first be evaluated and their return values passed.
            This is useful for runtime configuration.
    """
    success_hdlrs = _handlers(on_success)
    backoff_hdlrs = _handlers(on_backoff, _log_backoff)
    giveup_hdlrs = _handlers(on_giveup, _log_giveup)

    def decorate(target):

        @functools.wraps(target)
        def retry(*args, **kwargs):
            # change names because python 2.x doesn't have nonlocal
            max_tries_ = _maybe_call(max_tries)

            tries = 0
            wait = _init_wait_gen(wait_gen, wait_gen_kwargs)
            while True:
                tries += 1
                details = (target, args, kwargs, tries)

                ret = target(*args, **kwargs)
                if predicate(ret):
                    if tries == max_tries_:
                        _call_handlers(giveup_hdlrs, *details, value=ret)
                        break

                    seconds = _next_wait(wait, jitter)

                    _call_handlers(backoff_hdlrs, *details,
                                   value=ret, wait=seconds)

                    time.sleep(seconds)
                    continue
                else:
                    _call_handlers(success_hdlrs, *details, value=ret)
                    break

            return ret

        return retry

    # Return a function which decorates a target with a retry loop.
    return decorate


def on_exception(wait_gen,
                 exception,
                 max_tries=None,
                 jitter=full_jitter,
                 giveup=lambda e: False,
                 on_success=None,
                 on_backoff=None,
                 on_giveup=None,
                 **wait_gen_kwargs):
    """Returns decorator for backoff and retry triggered by exception.

    Args:
        wait_gen: A generator yielding successive wait times in
            seconds.
        exception: An exception type (or tuple of types) which triggers
            backoff.
        max_tries: The maximum number of attempts to make before giving
            up. Once exhausted, the exception will be allowed to escape.
            The default value of None means their is no limit to the
            number of tries. If a callable is passed, it will be
            evaluated at runtime and its return value used.
        jitter: A function of the value yielded by wait_gen returning
            the actual time to wait. This distributes wait times
            stochastically in order to avoid timing collisions across
            concurrent clients. Wait times are jittered by default
            using the full_jitter function. Jittering may be disabled
            altogether by passing jitter=None.
        giveup: Function accepting an exception instance and
            returning whether or not to give up. Optional. The default
            is to always continue.
        on_success: Callable (or iterable of callables) with a unary
            signature to be called in the event of success. The
            parameter is a dict containing details about the invocation.
        on_backoff: Callable (or iterable of callables) with a unary
            signature to be called in the event of a backoff. The
            parameter is a dict containing details about the invocation.
        on_giveup: Callable (or iterable of callables) with a unary
            signature to be called in the event that max_tries
            is exceeded.  The parameter is a dict containing details
            about the invocation.
        **wait_gen_kwargs: Any additional keyword args specified will be
            passed to wait_gen when it is initialized.  Any callable
            args will first be evaluated and their return values passed.
            This is useful for runtime configuration.
    """
    success_hdlrs = _handlers(on_success)
    backoff_hdlrs = _handlers(on_backoff, _log_backoff)
    giveup_hdlrs = _handlers(on_giveup, _log_giveup)

    def decorate(target):

        @functools.wraps(target)
        def retry(*args, **kwargs):
            # change names because python 2.x doesn't have nonlocal
            max_tries_ = _maybe_call(max_tries)

            tries = 0
            wait = _init_wait_gen(wait_gen, wait_gen_kwargs)
            while True:
                tries += 1
                details = (target, args, kwargs, tries)

                try:
                    ret = target(*args, **kwargs)
                except exception as e:
                    if giveup(e) or tries == max_tries_:
                        _call_handlers(giveup_hdlrs, *details)
                        raise

                    seconds = _next_wait(wait, jitter)

                    _call_handlers(backoff_hdlrs, *details, wait=seconds)

                    time.sleep(seconds)
                else:
                    _call_handlers(success_hdlrs, *details)

                    return ret

        return retry

    # Return a function which decorates a target with a retry loop.
    return decorate


# Evaluate arg that can be either a fixed value or a callable.
def _maybe_call(f, *args, **kwargs):
    return f(*args, **kwargs) if callable(f) else f


def _init_wait_gen(wait_gen, wait_gen_kwargs):
    # there are no dictionary comprehensions in python 2.6
    kwargs = dict((k, _maybe_call(v))
                  for k, v in wait_gen_kwargs.items())
    return wait_gen(**kwargs)


def _next_wait(wait, jitter):
    value = next(wait)
    try:
        if jitter is not None:
            seconds = jitter(value)
        else:
            seconds = value
    except TypeError:
        # support deprecated nullary jitter function signature
        # which returns a delta rather than a jittered value
        seconds = value + jitter()

    return seconds


def _call_handlers(hdlrs, target, args, kwargs, tries, **extra):
    details = {
        'target': target,
        'args': args,
        'kwargs': kwargs,
        'tries': tries,
    }
    details.update(extra)
    for hdlr in hdlrs:
        hdlr(details)


# Create default handler list from keyword argument
def _handlers(hdlr, default=None):
    defaults = [default] if default is not None else []

    if hdlr is None:
        return defaults

    if hasattr(hdlr, '__iter__'):
        return defaults + list(hdlr)

    return defaults + [hdlr]


# Default backoff handler
def _log_backoff(details):
    fmt = "Backing off {0}(...) for {1:.1f}s"
    msg = fmt.format(details['target'].__name__, details['wait'])

    exc_typ, exc, _ = sys.exc_info()
    if exc is not None:
        exc_fmt = traceback.format_exception_only(exc_typ, exc)[-1]
        msg = "{0} ({1})".format(msg, exc_fmt.rstrip("\n"))
        logger.error(msg)
    else:
        msg = "{0} ({1})".format(msg, details['value'])
        logger.info(msg)


# Default giveup handler
def _log_giveup(details):
    fmt = "Giving up {0}(...) after {1} tries"
    msg = fmt.format(details['target'].__name__, details['tries'])

    exc_typ, exc, _ = sys.exc_info()
    if exc is not None:
        exc_fmt = traceback.format_exception_only(exc_typ, exc)[-1]
        msg = "{0} ({1})".format(msg, exc_fmt.rstrip("\n"))
    else:
        msg = "{0} ({1})".format(msg, details['value'])

    logger.error(msg)
